### PR TITLE
Refactor OverlayedChanges

### PR DESCRIPTION
**Pull request by athei**
# Overview
This changes the visibility of `OverlayChanges`members from public to private. Users are migrated from directly accessing its data to using high level functions. Changes to the test code are factored out into a separate commit as these are noisy and distract from the core changes.

# Motivation
Currently, many parts of the `state-machine` crate are having assumptions about the internal structure of `OverlayChanges`baked in. This leaky abstraction prevents improvements to the `OverlayChanges`without triggering a huge code churn. One example would be implementing sotrage transactions which would change the internal layout and therefore brakes all users. This PR fixes this by removing those assumptions of the users. It is a prerequisite for implementing storage transactions in a way that does not result in a huge diff scattered between many files.

### CODE REVIEW

File: primitives/state-machine/src/overlayed_changes.rs [hunk 1]
```diff
@@ -498,6 +508,42 @@ impl OverlayedChanges {
...
+		committed.into_iter().flatten().chain(prospective.into_iter().flatten())
+	}
+
+	/// Return a clone of the currently pending changes.
+	pub fn clone_pending(&self) -> OverlayedChangeSet {
+		self.prospective.clone()
+	}
+
+	/// Replace the currently pending changes.
+	pub fn set_pending(&mut self, pending: OverlayedChangeSet) {
```
Comment (pepyakin) [inline primitives/state-machine/src/overlayed_changes.rs:543]: I wonder if it is a good idea to introduce another name for the prospective here. 

Do you think it would be changed or removed with the upcoming PR?

File: primitives/state-machine/src/overlayed_changes.rs [hunk 2]
```diff
@@ -498,6 +508,42 @@ impl OverlayedChanges {
...
+			.chain(self.committed.children_default.iter())
+			.map(|(_, v)| &v.1)
+	}
+
+	/// Get an iterator over all pending and committed changes.
+	///
+	/// Supplying `None` for `child_info` will only return changes that are in the top
+	/// tree. Specifying some `child_info` will return only the changes in that
+	/// child tree.
+	pub fn changes(&self, child_info: Option<&ChildInfo>) -> impl Iterator<Item=(&StorageKey, &OverlayedValue)> {
```
Comment (pepyakin) [inline primitives/state-machine/src/overlayed_changes.rs:523]: nit: line is unnecessarily longer than the soft limit of 100.

File: primitives/state-machine/src/overlayed_changes.rs [hunk 3]
```diff
@@ -498,6 +508,42 @@ impl OverlayedChanges {
 		)
 	}
 
+	/// Get an iterator over all child trees that are stored in the overlay.
```
Comment (pepyakin) [inline primitives/state-machine/src/overlayed_changes.rs:511]: typo: s/trees/tries

File: primitives/state-machine/src/overlayed_changes.rs [hunk 4]
```diff
@@ -498,6 +508,42 @@ impl OverlayedChanges {
...
+	pub fn child_infos(&self) -> impl Iterator<Item=&ChildInfo> {
+		self.prospective.children_default.iter()
+			.chain(self.committed.children_default.iter())
+			.map(|(_, v)| &v.1)
+	}
+
+	/// Get an iterator over all pending and committed changes.
+	///
+	/// Supplying `None` for `child_info` will only return changes that are in the top
+	/// tree. Specifying some `child_info` will return only the changes in that
```
Comment (pepyakin) [inline primitives/state-machine/src/overlayed_changes.rs:521]: typo: s/tree/trie

File: primitives/state-machine/src/overlayed_changes.rs [hunk 5]
```diff
@@ -186,6 +186,16 @@ impl FromIterator<(StorageKey, OverlayedValue)> for OverlayedChangeSet {
...
 
+impl OverlayedValue {
+	pub fn value(&self) -> Option<&StorageValue> {
+		self.value.as_ref()
+	}
+
+	pub fn extrinsics(&self) -> Option<impl Iterator<Item=&u32>> {
+		self.extrinsics.as_ref().map(|v| v.iter())
+	}
+}
```
Comment (pepyakin) [inline primitives/state-machine/src/overlayed_changes.rs:197]: Those two could use a doc comment, esp. `extrinsics`.

File: primitives/state-machine/src/overlayed_changes.rs [hunk 6]
```diff
@@ -498,6 +508,42 @@ impl OverlayedChanges {
...
+		let (committed, prospective) = if let Some(child_info) = child_info {
+			match child_info.child_type() {
+				ChildType::ParentKeyId => (
+					self.committed.children_default.get(child_info.storage_key()).map(|c| &c.0),
+					self.prospective.children_default.get(child_info.storage_key()).map(|c| &c.0),
+				),
+			}
+		} else {
+			(Some(&self.committed.top), Some(&self.prospective.top))
+		};
```
Comment (pepyakin) [inline primitives/state-machine/src/overlayed_changes.rs:533]: nit: the single `Option` parameter which changes the behavior of this function significantly suggest that maybe we should separate this function on two? A drawback that I see are minor code duplication for the line below which IMO. But doc and logic would be a bit simpler and the call sites would be a bit tider.

File: primitives/state-machine/src/ext.rs [hunk 7]
```diff
@@ -745,7 +727,7 @@ mod tests {
 		let mut ext = TestExt::new(&mut overlay, &mut offchain_overlay, &mut cache, &backend, state, None);
 		assert_eq!(
 			ext.storage_changes_root(&H256::default().encode()).unwrap(),
-			Some(hex!("bb0c2ef6e1d36d5490f9766cfcc7dfe2a6ca804504c3bb206053890d6dd02376").to_vec()),
+			Some(hex!("dfb314e4d5ddfaed1b6ef8d98d9898971bffb303a5d50e0b9c9a7bfd42ccdb90").to_vec()),
```
Comment (pepyakin) [inline primitives/state-machine/src/ext.rs:730]: This is a bit worrisome because I expected those changes to be pretty cosmetical.

Why was this changed?

Comment (athei) [reply primitives/state-machine/src/ext.rs:730]: It only changed because with the public functions the exact overlay layout cannot be reproduced. It changed slightly because we cannot set the modifying extrinsic of the `EXTRINSIC_INDEX` entry itself (nor it does make sense to do so). I verified that the root is the same when I directly set the fields. So it is only caused by the changes to the tests.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 1 above)
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs:543]: You successfully identified the remaining leak in the abstraction where we still have assumptions. I consider it a place holder that goes away with the upcoming PR. It will be replaced by rolling back a transaction. I guess even now we can replace it by just calling`discard_prospective` as the `prospective` should be empty as it is committed for every call. Actually no need to clone it. I will try if that works. Then we do not need to add further functions in that PR.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 2 above)
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs:523]: Correct. I will change it.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 6 above)
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs:533]: As a matter of fact I chose this approach because it simplifies the calling site where the code is the same for child trees and top trees. Splitting that up would require the caller to chain top and child trie changes together.

That said, I actually think splitting would be the right thing to do I will investigate.

Comment (pepyakin) [reply primitives/state-machine/src/overlayed_changes.rs:533]: Ah, I didn't account for that fact. Up to you then!

File: primitives/state-machine/src/ext.rs (hunk 7 above)
Comment (pepyakin) [reply primitives/state-machine/src/ext.rs:730]: Ok, sounds plausible.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 1 above)
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs]: I removed the cloning completely. It appears that just resetting the `prospective` works just as fine. I removed these two new functions. Eventually this `discard_prospective` goes away and is replaced by `rollback_transaction`.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 6 above)
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs:537]: Put some more thought into it. I cannot really come up with something better that satisfies every user. I will leave it as is in absence of further arguments.

File: primitives/state-machine/src/overlayed_changes.rs [hunk 8]
```diff
@@ -498,6 +510,34 @@ impl OverlayedChanges {
 		)
 	}
 
+	/// Get an iterator over all child tries that are stored in the overlay.
```
Comment (bkchr) [inline primitives/state-machine/src/overlayed_changes.rs:513]:
```suggestion
	/// Get an iterator over all pending and committed child tries in the overlay.
	///
	/// NOTE: The returned iterator can return the same child trie multiple times.
```

File: primitives/state-machine/src/changes_trie/build.rs [hunk 9]
```diff
@@ -105,20 +105,18 @@ fn prepare_extrinsics_input<'a, B, H, Number>(
...
-
-	let mut children_info = BTreeSet::<ChildInfo>::new();
 	let mut children_result = BTreeMap::new();
-	for (_storage_key, (_map, child_info)) in changes.prospective.children_default.iter()
-		.chain(changes.committed.children_default.iter()) {
-		children_info.insert(child_info.clone());
-	}
-	for child_info in children_info {
+
+	for child_info in changes.child_infos() {
```
Comment (bkchr) [inline primitives/state-machine/src/changes_trie/build.rs:110]: This is wrong. Before we filtered duplicates by using the `BTreeSet` and now we don't filter the duplicates. It may not change the output (as far as I can see and understand the code on a short review), however it will probably be faster to filter duplicates before iterating them.

File: primitives/state-machine/src/lib.rs [hunk 10]
```diff
@@ -349,7 +347,7 @@ impl<'a, B, H, N, Exec> StateMachine<'a, B, H, N, Exec> where
 		let (result, was_native) = self.execute_aux(true, native_call.take());
 
 		if was_native {
-			self.overlay.prospective = orig_prospective.clone();
+			self.overlay.discard_prospective();
```
Comment (bkchr) [inline primitives/state-machine/src/lib.rs:350]: This is wrong, we need to restore the other prospective changes.

File: primitives/state-machine/src/lib.rs [hunk 11]
```diff
@@ -385,7 +382,7 @@ impl<'a, B, H, N, Exec> StateMachine<'a, B, H, N, Exec> where
 		if !was_native || result.is_ok() {
 			result
 		} else {
-			self.overlay.prospective = orig_prospective.clone();
+			self.overlay.discard_prospective();
```
Comment (bkchr) [inline primitives/state-machine/src/lib.rs:385]: Same.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 1 above)
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs:547]: After talking to @bkchr we came to the conclusion that we cannot make this assumption. I revert the commit. We are back to the `pending_changes()` function until storage transaction come around.

File: primitives/state-machine/src/lib.rs (hunk 10 above)
Comment (athei) [reply primitives/state-machine/src/lib.rs]: I reverted the change.

File: primitives/state-machine/src/lib.rs (hunk 11 above)
Comment (athei) [reply primitives/state-machine/src/lib.rs]: I reverted the change.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 8 above)
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs:513]: Shouldn't we rather just de duplicate the child infos here in this function?

Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs]: I opted for de de duplicating them in the `fn child_infos()` as the duplicates would go away with the storage transaction changes anyways.

File: primitives/state-machine/src/changes_trie/build.rs (hunk 9 above)
Comment (athei) [reply primitives/state-machine/src/changes_trie/build.rs:110]: Added de duplication in the `fn child_infos`.

**Comment by athei** [comment]
All issues are addresses and successfully synced kusama with the `Both`strategy from scratch.

**Comment by bkchr** [comment]
> All issues are addressed and I successfully synced kusama with the `Both`strategy from scratch.

Could you give your branch please to @ddorgan to test it on a validator? (I want to be extra sure that we did not fuck up anything :sweat_smile:)

File: primitives/state-machine/src/overlayed_changes.rs [hunk 12]
```diff
@@ -498,6 +510,44 @@ impl OverlayedChanges {
...
+			.chain(self.committed.children_default.iter())
+			.map(|(_, v)| &v.1).collect::<BTreeSet<_>>()
+	}
+
+	/// Get an iterator over all pending and committed changes.
+	///
+	/// Supplying `None` for `child_info` will only return changes that are in the top
+	/// trie. Specifying some `child_info` will return only the changes in that
+	/// child trie.
+	pub fn changes(&self, child_info: Option<&ChildInfo>)
```
Comment (cheme) [inline primitives/state-machine/src/overlayed_changes.rs:525]: I like the idea of an api using `Option<childinfo>` (got similar code in some variant of my prs), note that you could also remove option and use child_info.is_top_trie (not sure if the fonction currently existr but it is in many of my prs).
Then you got choice to define top trie as a separate child type, but in my previous pr I just did define it as the default child trie located at empty key location (it removes the ability to define a default child trie at such location, but such child trie key is invalid for a default child trie anyway).

File: primitives/state-machine/src/overlayed_changes.rs [hunk 13]
```diff
@@ -498,6 +510,44 @@ impl OverlayedChanges {
...
+		committed.into_iter().flatten().chain(prospective.into_iter().flatten())
+	}
+
+	/// Return a clone of the currently pending changes.
+	pub fn clone_pending(&self) -> OverlayedChangeSet {
+		self.prospective.clone()
+	}
+
+	/// Replace the currently pending changes.
+	pub fn set_pending(&mut self, pending: OverlayedChangeSet) {
```
Comment (cheme) [inline primitives/state-machine/src/overlayed_changes.rs:547]: the function erase pending change, so `replace_pending`, `emplace_pending` or similar name could also be possible.

File: primitives/state-machine/src/testing.rs [hunk 14]
```diff
@@ -135,21 +135,19 @@ impl<H: Hasher, N: ChangesTrieBlockNumber> TestExternalities<H, N>
...
-			.chain(self.overlay.prospective.children_default.clone().into_iter())
-			.for_each(|(_storage_key, (map, child_info))| {
-				transaction.push((
-					Some(child_info),
-					map.into_iter()
-						.map(|(k, v)| (k, v.value))
-						.collect::<Vec<_>>(),
-				))
-			});
+		for child_info in self.overlay.child_infos() {
```
Comment (cheme) [inline primitives/state-machine/src/testing.rs:143]: Here child_infos is currently building an unnecessary btreeset.
Also we are iterating child infos first in `child_infos` function, then we access the corresponding values a second time in `changes` when we could have use them from the first iteration.
Depending on what should be the next steps I wouldn't spend to much time on this (just avoid the internal btreeset).
Different api could be use, I just check I did use `chidren_iter_overlay` function in #3263, IIrc it is probably simply returning an double iterator.
But truely at this point we should not expect huge number of items in child trie.

File: primitives/state-machine/src/lib.rs [hunk 15]
```diff
@@ -1005,7 +994,8 @@ mod tests {
 		overlay.commit_prospective();
 
 		assert_eq!(
-			overlay.committed,
+			overlay.changes(None).map(|(k, v)| (k.clone(), v.value().cloned()))
+				.collect::<HashMap<_, _>>(),
```
Comment (cheme) [inline primitives/state-machine/src/lib.rs:998]: This is not strictly testing the same thing, the previous test was asserting that the effect of `commit_prospective`.
An easy way to do it would be to call `overlay.set_pending(Default::default)` before the assertion.

File: primitives/state-machine/src/ext.rs [hunk 16]
```diff
@@ -754,14 +736,14 @@ mod tests {
 		let mut overlay = prepare_overlay_with_changes();
 		let mut offchain_overlay = prepare_offchain_overlay_with_changes();
 		let mut cache = StorageTransactionCache::default();
-		overlay.prospective.top.get_mut(&vec![1]).unwrap().value = None;
+		overlay.set_storage(vec![1], None);
```
Comment (cheme) [inline primitives/state-machine/src/ext.rs:739]: This break the test, the test was testing that storage_changes_root_is_some_when_extrinsic_changes_are_empty , fwiu not registering the extrinsic change was intentional here.
Maybe this can be achieved by changes.set_collect_extrinsics(false); then set_storage then changes.set_collect_extrinsics(true);.
The root should not change.

File: primitives/state-machine/src/changes_trie/build.rs [hunk 17]
```diff
@@ -105,20 +105,18 @@ fn prepare_extrinsics_input<'a, B, H, Number>(
...
 		let child_index = ChildIndex::<Number> {
 			block: block.clone(),
 			storage_key: child_info.prefixed_storage_key(),
 		};
 
-		let iter = prepare_extrinsics_input_inner(backend, block, changes, Some(child_info))?;
+		let iter = prepare_extrinsics_input_inner(
+			backend, block, changes,
```
Comment (cheme) [inline primitives/state-machine/src/changes_trie/build.rs:117]: I really have no idea if it is ok formatting to mix arguments in a same line with arguments in multiple line, just mentioning it.

File: primitives/state-machine/src/ext.rs [hunk 18]
```diff
@@ -669,35 +662,24 @@ mod tests {
...
-					extrinsics: Some(vec![1].into_iter().collect())
-				}),
-			].into_iter().collect(),
-			committed: Default::default(),
-			collect_extrinsics: true,
-			stats: Default::default(),
-		}
+		let mut changes = OverlayedChanges::default();
+		changes.set_collect_extrinsics(true);
+		changes.set_extrinsic_index(3);
```
Comment (cheme) [inline primitives/state-machine/src/ext.rs:680]: why not set_extrinsic index(1) and set 3 later to have same roots?

File: primitives/state-machine/src/changes_trie/build.rs [hunk 19]
```diff
@@ -666,10 +629,7 @@ mod test {
 			let (backend, storage, mut changes, config) = prepare_for_build(zero);
 
 			// 110: missing from backend, set to None in overlay
-			changes.prospective.top.insert(vec![110], OverlayedValue {
-				value: None,
-				extrinsics: Some(vec![1].into_iter().collect())
-			});
+			changes.set_storage(vec![110], None);
```
Comment (cheme) [inline primitives/state-machine/src/changes_trie/build.rs:632]: set_extirnsic_index(1) before? (and restore afterward)

File: primitives/state-machine/src/overlayed_changes.rs (hunk 12 above)
Comment (cheme) [reply primitives/state-machine/src/overlayed_changes.rs:525]: Ok, please forget this comment, I realize that using Option is something I probably have wrote at some point to avoid making top trie specifically a child trie, it is out of scope of this pr.

File: primitives/state-machine/src/testing.rs (hunk 14 above)
Comment (athei) [reply primitives/state-machine/src/testing.rs:143]: Yes this is doing work that is strictly not necessary by going over the children multiple times. However, as this is only test code I opted for not adding an additional API just for the sake of some potential speedup that does not even matter.

File: primitives/state-machine/src/overlayed_changes.rs [hunk 20]
```diff
@@ -498,6 +510,44 @@ impl OverlayedChanges {
 		)
 	}
 
+	/// Get an iterator over all pending and committed child tries in the overlay.
+	pub fn child_infos(&self) -> impl IntoIterator<Item=&ChildInfo> {
+		self.prospective.children_default.iter()
+			.chain(self.committed.children_default.iter())
+			.map(|(_, v)| &v.1).collect::<BTreeSet<_>>()
```
[the remark this answers is not in the document]
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs:517]: I like the idea of swapping them for consistency. However, I won't be optimizing it further with some iterator as this code is going away with storage tx when there is only one map.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 13 above)
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs:547]: Yes that makes sense. However that pending notion will also go away soon and will just be replaced by a transaction.

File: primitives/state-machine/src/lib.rs (hunk 15 above)
Comment (athei) [reply primitives/state-machine/src/lib.rs:998]: Isn't the `commit_prospective` above the assert achieving the same effect?

File: primitives/state-machine/src/testing.rs (hunk 14 above)
Comment (cheme) [reply primitives/state-machine/src/testing.rs:143]: Ah, we could think about making `child_infos` a test function then?

File: primitives/state-machine/src/overlayed_changes.rs (hunk 13 above)
Comment (cheme) [reply primitives/state-machine/src/overlayed_changes.rs:547]: ok, then it is probably best to use old `prospective` name until future changes.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (cheme) [reply primitives/state-machine/src/overlayed_changes.rs:517]: when will this go away? I am asking because when working on #3263 I got conflicting requirement (I did dumb down the code at some point but it introduces O(N), and https://github.com/paritytech/substrate/pull/3263#issuecomment-573559259 after I switch back to the old version (probably on of the reason the PR refactor did not reach consensus)).
Just not sure if it will be quick to put in place.

File: primitives/state-machine/src/lib.rs (hunk 15 above)
Comment (cheme) [reply primitives/state-machine/src/lib.rs:998]: commit prospective is merging both layer, the test assert that all the resulting values are in the bottom layer.
So your new version does not test this anymore, dropping the prospective layer will again test this.

File: primitives/state-machine/src/testing.rs (hunk 14 above)
Comment (athei) [reply primitives/state-machine/src/testing.rs:143]: No it is also used in production code.

File: primitives/state-machine/src/lib.rs (hunk 15 above)
Comment (athei) [reply primitives/state-machine/src/lib.rs:998]: The commit was already done in the original test. Just dropping it here would brake the test.

**Comment by gnunicorn** [comment]
waiting for @ddorgan to confirm this works on a validator node without hiccups.

File: primitives/state-machine/src/overlayed_changes.rs (hunk 13 above)
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs]: I renamed it to `replace_pending`. I feel like the prospective name is referring strictly to a specific internal layout when other goal is to make the interface more abstract.

File: primitives/state-machine/src/ext.rs (hunk 16 above)
Comment (athei) [reply primitives/state-machine/src/ext.rs:742]: Good point. Done.

File: primitives/state-machine/src/changes_trie/build.rs (hunk 17 above)
Comment (athei) [reply primitives/state-machine/src/changes_trie/build.rs:117]: Me neither. Let's just leave it be until someone with a strong opinion comes around.

File: primitives/state-machine/src/changes_trie/build.rs (hunk 19 above)
Comment (athei) [reply primitives/state-machine/src/changes_trie/build.rs:634]: I made it so that the extrinsic index is set to 1 now (like before). Not that it matters.

File: primitives/state-machine/src/ext.rs (hunk 18 above)
Comment (athei) [reply primitives/state-machine/src/ext.rs]: Good point. I reordered the statements so that we no longer have to change the storage root.

File: primitives/state-machine/src/lib.rs (hunk 15 above)
Comment (cheme) [reply primitives/state-machine/src/lib.rs:998]: I did not mean drop the commit, but drop the prospective.

But yes that is probably not the original intention of the test (thus the 'strictly' in my original comment), so it does not matter that much :)

File: primitives/state-machine/src/overlayed_changes.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (cheme) [reply primitives/state-machine/src/overlayed_changes.rs:517]:
```rust
	/// Get an iterator over all pending and committed child tries in the overlay.
	pub fn child_infos(&self) -> impl IntoIterator<Item=&ChildInfo> {
		use std::cmp::Ordering;
		let mut committed = self.committed.children_default.iter()
			.map(|(_, v)| &v.1);
		let mut prospective = self.prospective.children_default.iter()
			.map(|(_, v)| &v.1);
		let mut committed_next = committed.next();
		let mut prospective_next = prospective.next();
		std::iter::from_fn(move || {
			match committed_next.cmp(&prospective_next) {
				Ordering::Less => mem::replace(&mut committed_next, committed.next()),
				Ordering::Greater => mem::replace(&mut prospective_next, prospective.next()),
				Ordering::Equal => {
					committed_next = committed.next();
					mem::replace(&mut prospective_next, prospective.next())
				},
			}
		})
	}
```
or anything similar ?
but it is only a nitpick.

[the remark this answers is not in the document]
Comment (athei) [reply primitives/state-machine/src/overlayed_changes.rs:517]: Thing is that my change is not worsening the situation as it was building a `BTreeSet` before. When implementing something that is too complicated to verify that it is correct just by looking at it we need a test for that. With the absence of that I rather stick with the simple solution.

I appreciate your effort though.

[the remark this answers is not in the document]
Comment (cheme) [reply primitives/state-machine/src/overlayed_changes.rs:517]: Np, I did not realize it was the existing code, actually this snippert that is probably some code I already did write many time in many of my unmerged pr(s).

**Comment by gavofyork** [comment]
@ddorgan can you confirm this?

**Comment by athei** [comment]
Rebased onto current master so it works with the current polkadot master, again.

**Comment by ddorgan** [comment]
@gavofyork looks good on two validators (one on flaming fir and one on kusama) after 24 hours.
